feat(dashboards): Add release `SearchQueryBuilder`

gggritso opened the pull request:
For the release dataset, if the `'search-query-builder-releases'` is on, show a `SearchQueryBuilder` instead of a regular search bar. Note: This does _not_ use `'search-query-builder-dashboards'` flag by design. This follows how the logic works for `'search-query-builder-issues'`, which, if enabled, changes the issue search bar _everywhere_ including dashboards.

**Before:**
<img width="957" alt="Screenshot 2024-09-04 at 2 13 11 PM" src="https://github.com/user-attachments/assets/6fd84671-f955-47c7-9f3a-b9403cf88a37">

**After:**  
<img width="713" alt="Screenshot 2024-09-04 at 2 24 43 PM" src="https://github.com/user-attachments/assets/b8c761cd-b14f-42a8-807b-13b4a4552374">

gggritso committed:
Add query builder conditional

If the flag is on, render a `SearchQueryBuilder` instead of a regular
search bar.

malwilley commented on static/app/views/dashboards/widgetBuilder/buildSteps/filterResultsStep/releaseSearchBar.tsx:
```diff
@@ -75,7 +81,26 @@ export function ReleaseSearchBar({
+    <SearchQueryBuilder
+      initialQuery={widgetQuery.conditions}
+      filterKeySections={filterKeySections}
+      filterKeys={supportedTags}
```
Could you extract `supportedTags` outside of the component so that the reference doesn't change on rerender? The new component is pretty heavy so we want to avoid it where possible

malwilley commented on static/app/views/dashboards/widgetBuilder/buildSteps/filterResultsStep/releaseSearchBar.tsx:
```diff
@@ -75,7 +81,26 @@ export function ReleaseSearchBar({
+      disallowUnsupportedFilters
+      disallowFreeText
+      recentSearches={SavedSearchType.SESSION}
+    />
```
Do you need to pass the `invalidMessages` here? The previous version had that and we still support that prop

gggritso replied:
I thought the default messages were pretty good, but I added explicit ones 👍🏻

malwilley commented on static/app/views/dashboards/widgetBuilder/buildSteps/filterResultsStep/releaseSearchBar.tsx:
```diff
@@ -75,7 +81,26 @@ export function ReleaseSearchBar({
+      initialQuery={widgetQuery.conditions}
+      filterKeySections={filterKeySections}
+      filterKeys={supportedTags}
+      getTagValues={memoize(
```
No need to wrap in `memoize` since the component will deal with caching for you now

gggritso replied:
I thought the default messages were pretty good, but I added explicit ones 👍🏻

gggritso committed:
Lift up `supportedTags` definition

This prevents its identity changing on every render and forcing a
re-render of the query builder.

gggritso committed:
Set explicit invalid messages

gggritso committed:
Remove unneeded memoize
